### PR TITLE
import onnx ml operations as a separate dialect

### PR DESCRIPTION
**Pull request by chentong319**
1. gen_doc.py
A option is added to generate inc files for onnx ml operation: --domain="ONNX_ML". Default is ONNX for onnx dialect

2. MLONNX dialect
In Dialect/MLONNX, basic operation definition with TableGen.

3. Builder
 MLONNX is included in the frontend.

4. Project control
By default, the MLONNX dialect is included in build. Use EXCLUDE_ONNX_ML=1 to cmake to disable  ONNX ML in build. 

After this PR is merged, other components for MLONNX, such as documentation and test, will be added.

### CODE REVIEW

**Comment by doru1004** [comment]
> 1. gen_doc.py
>    A option is added to generate inc files for onnx ml operation: --domain="ONNX_ML". Default is ONNX for onnx dialect

Can we refactor this into separate generator files? It would be much cleaner and easier to maintain if we had a generator file for each dialect. We should also take this opportunity to change the name to something more meaningful:

```
gen-onnx.py
gen-onnx-ml.py
```

Any functionality that is common to the two generators can be factored out into a separate file:

```
gen_utils.py
```

**Comment by chentong319** [comment]
The two share all the code except for the name of dialect in output. No need to separate them. Unnecessary effort at least at current stage.

File: src/Dialect/MLONNX/MLONNXOps.td [hunk 1]
```diff
@@ -0,0 +1,67 @@
+//===--- MLONNXOps.td -- ONNX ML Dialect Operation Definitions ----*- tablegen -===//
```
Comment (doru1004) [inline src/Dialect/MLONNX/MLONNXOps.td:1]: Line too long.

File: src/Dialect/CMakeLists.txt [hunk 2]
```diff
@@ -1,2 +1,5 @@
 add_subdirectory(Krnl)
-add_subdirectory(ONNX)
\ No newline at end of file
+add_subdirectory(ONNX)
+if (INCLUDE_ONNX_ML)
```
Comment (doru1004) [inline src/Dialect/CMakeLists.txt:3]: I think we should be building ONNX ML unconditionally. It's one of the dialects we supports. MLIR builds all its dialects and then it's up to the user which ones get registered and used.

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -43,6 +43,11 @@ target_link_libraries(onnx-mlir
         ${MLIRLibs}
         ${CMAKE_DL_LIBS})
 
+if (INCLUDE_ONNX_ML)
```
Comment (doru1004) [inline src/CMakeLists.txt:46]: Always build it.

**Comment by doru1004** [comment]
I think we need to always generate and build the ONNX ML since it's a dialect that we support. Using the dialect is up to the user.

File: src/CMakeLists.txt (hunk 3 above)
Comment (chentong319) [reply src/CMakeLists.txt:46]: By default, we build the onnx ml. At a meeting, someone asked for the support that we choose not to build onnx ml.  I have no preference over this issue. @tjingrant  @AlexandreEichenberger What's your opinion?

File: src/Dialect/MLONNX/MLONNXOps.td [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+//
+// Copyright 2019-2020 The IBM Research Authors
+//
+// =============================================================================
+//
+// Defines ONNX Dialect operations.
```
[the remark this answers is not in the document]
Comment (chentong319) [reply src/Dialect/MLONNX/MLONNXOps.td]: All the format issues are fixed.

**Comment by s390xlinux** [comment]
Can one of the admins verify this patch?

**Comment by chentong319** [comment]
@jenkins-droid test this please.

**Comment by chentong319** [comment]
@doru1004 I updated this patch and invoked the Jenkins. Can we complete this patch first?

**Pull request merged**